src/peergos/shared/social/SocialFeed.java

```diff
@@ -0,0 +1,285 @@
+                                diff.updatedWriteBytes());
+                    }
+                    // There must have been concurrent processing of new caps, e.g. by browsing to that user's files
+                    // We need to load the whole lot to catch what we missed
+                    return friend.loadCachedCaps(context.network, context.crypto)
+                            .thenCompose(all -> {
+                                List<CapabilityWithPath> readCapsToAdd = Stream.of(all.readCaps, diff.newCaps.readCaps)
+                                        .flatMap(c -> c.getRetrievedCapabilities().stream())
+                                        .skip(current.readCaps)
+                                        .collect(Collectors.toList());
+                                List<CapabilityWithPath> writeCapsToAdd = Stream.of(all.writeCaps, diff.newCaps.writeCaps)
+                                        .flatMap(c -> c.getRetrievedCapabilities().stream())
+                                        .skip(current.writeCaps)
+                                        .collect(Collectors.toList());
+                                return addToFriend(friend.ownerName, current, readCapsToAdd, all.readCaps.getBytesRead(),
+                                        writeCapsToAdd, all.writeCaps.getBytesRead());
+                            });
+                });
+    }
+
+    private static String extractOwner(String path) {
+        int start = path.startsWith("/") ? 1 : 0;
+        int end = path.indexOf("/", start + 1);
+        return path.substring(start, end);
+    }
+
+    private synchronized CompletableFuture<SocialFeed> addToFriend(String friendName,
+                                                                   UserState current,
+                                                                   List<CapabilityWithPath> readCapsToAdd,
+                                                                   long updatedReadCapBytesTotal,
+                                                                   List<CapabilityWithPath> writeCapsToAdd,
+                                                                   long updateWriteCapBytesTotal) {
+        UserState updated = new UserState(
+                current.readCaps + readCapsToAdd.size(),
+                current.writeCaps + writeCapsToAdd.size(),
+                updatedReadCapBytesTotal,
+                updateWriteCapBytesTotal);
+        currentCapBytesProcessed.put(friendName, updated);
+        feedSizeRecords += readCapsToAdd.size() + writeCapsToAdd.size();
+        List<SharedItem> forFeed = Stream.of(readCapsToAdd, writeCapsToAdd)
+                .flatMap(List::stream)
+                .map(c -> new SharedItem(c.cap, extractOwner(c.path), friendName, c.path))
+                .collect(Collectors.toList());
+        ByteArrayOutputStream bout = new ByteArrayOutputStream();
+        for (SharedItem item : forFeed) {
```

Comment:
why not a stream rather than a loop?

Reply:
I don't like try catches inside streams - a loop is easier to read then I think.